### PR TITLE
Unmanaged GCE instances TCP health-checker

### PR DESCRIPTION
**Pull request by averbuks**
TCP healthchecker for unmanaged GCE instances based on Cloud Functions and VPC Serverless Access.
A good showcase for:
- VPC Serverless Access
- Cloud Scheduler + Cloud Functions
- Serverless for GCE operations

### CODE REVIEW

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+Stop `nginx` service on the test instance
+```
+gcloud compute ssh --zone europe-west1-b  nginx-test -- 'sudo systemctl stop nginx'
+```
+
+Wait a few minutes to allow scheduler to execute another healthcheck and examine the function logs
+```bash
+gcloud functions logs read cf-healthchecker --region=europe-west1
+
+#cf-healthchecker  ywn0bmojtrji  2022-03-15 21:59:21.202  Instanse restart task has been sent for instance nginx-test
```
Comment (ludoo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/README.md:59]: typo :)

**Review by ludoo — APPROVED** [review]
Hey Aleks, really nice. Could you add a sentence to the README to explain what kind of configuration is available for the health checker executable and CF?

I assume one could also run just the health checker and use its logs as a source for a custom metric and alert, right?

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md (hunk 1 above)
Comment (averbuks) [reply examples/cloud-operations/unmanaged-instances-healthcheck/README.md]: good catch, fixed

**Comment by averbuks** [comment]
> as a source for a custom metric and alert, right?

Added a bit more on the possible configuration options.
Speaking of alerts, yes they can be added with metric, but better to use `restarter` PubSub topic as a source for alerting and visualising how many instances failed over time.

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# TCP healthcheck and restart for unmanaged GCE instances
+
+This example shows how to leverage [Serverless VPC Access](https://cloud.google.com/vpc/docs/configure-serverless-vpc-access) and Cloud Functions to organize a highly performant TCP healtheck for unmanaged GCE instances. Healthchecker Cloud Function uses [goroutines](https://gobyexample.com/goroutines) to achieve parallel healthchecking for multiple instances and handles up to 1 thousand VMs checked in less than a second execution time.
```
Comment (rosmo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/README.md:3]: Typo: healtheck -> healthcheck.

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+- `FILTER` to filter list of GCE instances the health check will be targeted to. For instance `(name = nginx-*) AND (labels.env = dev)`
+- `GRACE_PERIOD` time period to prevent instance check of newly created instanced allowing services to start on the instance. 
+- `MAX_PARALLELISM` - max amount of healthchecks performed in parallel, be aware that every check requires an open TCP connection which is limited.
+- `PUBSUB_TOPIC` topic to publish the message with instance metadata. 
+- `RECHECK_INTERVAL` time period for performing recheck, when a check is failed it will be rechecked before marking as unhealthy. 
+- `TCP_PORT` port used for health checking
+- `TIMEOUT` the timeout time of a TCP probe.
+
+**_NOTE:_** In the current example `healthchecker` is used along with the `restarter` cloud function, but restarter can be replaced with another function like [Pubsub2Mailbox](https://github.com/GoogleCloudPlatform/professional-services/tree/main/tools/pubsub2inbox) for email notifications.
```
Comment (rosmo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/README.md:31]: Pubsub2Mailbox -> Pubsub2Inbox ;)

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+| [project_id](variables.tf#L33) | Project id to create a project when `project_create` is `true`, or to be used when `false`. | <code>string</code> | ✓ |  |
+| [grace_period](variables.tf#L56) | Grace period for an instance startup. | <code>string</code> |  | <code>&#34;180s&#34;</code> |
+| [location](variables.tf#L21) | App Engine location used in the example (required for CloudFunctions). | <code>string</code> |  | <code>&#34;europe-west&#34;</code> |
+| [project_create](variables.tf#L27) | Create project instead of using an existing one. | <code>bool</code> |  | <code>false</code> |
+| [region](variables.tf#L38) | Compute region used in the example. | <code>string</code> |  | <code>&#34;europe-west1&#34;</code> |
+| [root_node](variables.tf#L44) | The resource name of the parent folder or organization for project creation, in 'folders/folder_id' or 'organizations/org_id' format. | <code>string</code> |  | <code>null</code> |
+| [schedule](variables.tf#L50) | Cron schedule for executing compute instances healthckeck. | <code>string</code> |  | <code>&#34;&#42;&#47;5 &#42; &#42; &#42; &#42;&#34; &#35; every five minutes&#34;</code> |
```
Comment (rosmo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/README.md:108]: healthckeck -> healthcheck

File: examples/cloud-operations/unmanaged-instances-healthcheck/function/healthchecker/healthchecker.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+	return nil
+}
+
+// Start - Start Healthchecker
+func (hc *HealthChecker) Start() error {
+
+	wg := sync.WaitGroup{}
+
+	// Limit max parraller goroutines to prevent exosting socket connections
```
Comment (rosmo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/function/healthchecker/healthchecker.go:66]: exosting -> existing, parrallel -> parallel

File: examples/cloud-operations/unmanaged-instances-healthcheck/function/healthchecker/healthchecker.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	guard := make(chan struct{}, hc.config.max_parallelism)
```
Comment (rosmo) [inline examples/cloud-operations/unmanaged-instances-healthcheck/function/healthchecker/healthchecker.go:67]: Just thinking out loud here, but I wonder if `GOMAXPROCS` has an effect here (I think probably not as most time is spent blocking on a syscall).

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md (hunk 3 above)
Comment (averbuks) [reply examples/cloud-operations/unmanaged-instances-healthcheck/README.md:31]: ohhhh, terribly sorry, it was 2am;)) will fix shortly

File: examples/cloud-operations/unmanaged-instances-healthcheck/README.md (hunk 4 above)
Comment (averbuks) [reply examples/cloud-operations/unmanaged-instances-healthcheck/README.md]: done

**Pull request merged**